napari/_vispy/layers/labels.py

```diff
@@ -164,11 +174,46 @@ def hash2d_set(key, value, keys, values, empty_val=0):
             raise OverflowError('too many labels')
         key += 1
         pos = idx_to_2D(key, keys.shape)
-    keys[pos] = key
+    keys[pos] = initial_key
     values[pos] = value
 
 
-def build_textures_from_dict(color_dict, empty_val=0, shape=(1000, 1000)):
+def _get_shape_from_dict(color_dict):
+    size = len(color_dict) * 4
+    # I think that hash table size should be at least 4 times
+    # bigger than the number of labels to avoid collisions
```

Comment:
Should we have a MAX_LOAD_FACTOR constant, and maybe the tests should test the load factor rather than specific shapes? It would make it easier to run some experiments and revisit this decision in the future...

Reply:
The load factor here is `0.25`. I'm not sure how to understand next part.

Reply:
I mean it's nicer when the code refers to a named constant rather than `4`.

Additionally, the tests check that we return a specific shape. But that shape will change if we change the default load factor in the future. Instead, we should check that the load of the table is between 1/4 and 1/16. (ie between LOAD_FACTOR and LOAD_FACTOR / EXPANSION_FACTOR, which in this case is the same.)

Reply:
I will think. It is quite easy to test if the shader is enough big, but I'm not sure how to simply test if it is the smallest possible shader.